sdk/core/azure-core/test/ut/curl_session_test.cpp

```diff
@@ -82,6 +82,130 @@ namespace Azure { namespace Core { namespace Test {
+    // unique ptr for it
+    MockCurlNetworkConnection* curlMock = new MockCurlNetworkConnection();
+    EXPECT_CALL(*curlMock, SendBuffer(_, _, _)).WillOnce(Return(CURLE_OK));
+    EXPECT_CALL(*curlMock, ReadFromSocket(_, _, _))
+        .WillOnce(DoAll(
+            SetArrayArgument<1>(response.data(), response.data() + response.size()),
+            Return(response.size())))
+        .WillOnce(DoAll(
+            SetArrayArgument<1>(response2.data(), response2.data() + response2.size()),
+            Return(response2.size())));
+    EXPECT_CALL(*curlMock, GetConnectionKey()).WillRepeatedly(ReturnRef(connectionKey));
+    EXPECT_CALL(*curlMock, updateLastUsageTime());
+    EXPECT_CALL(*curlMock, DestructObj());
+
+    // Create the unique ptr to take care about memory free at the end
+    std::unique_ptr<MockCurlNetworkConnection> uniqueCurlMock(curlMock);
+
+    // Simulate a request to be sent
+    Azure::Core::Http::Url url("http://microsoft.com");
+    Azure::Core::Http::Request request(Azure::Core::Http::HttpMethod::Get, url);
+
+    {
+      // Create the session inside scope so it is released and the connection is moved to the pool
+      auto session = std::make_unique<Azure::Core::Http::CurlSession>(
+          request, std::move(uniqueCurlMock), true);
+
+      EXPECT_NO_THROW(session->Perform(Azure::Core::GetApplicationContext()));
+      auto response = session->GetResponse();
+      response->SetBodyStream(std::move(session));
+      auto bodyS = response->GetBodyStream();
+
+      // Read the bodyStream to get get all chunks
+      EXPECT_THROW(
+          Azure::Core::Http::BodyStream::ReadToEnd(Azure::Core::GetApplicationContext(), *bodyS),
+          Azure::Core::Http::TransportException);
+    }
+    // Clear the connections from the pool to invoke clean routine
+    Azure::Core::Http::CurlConnectionPool::ConnectionPoolIndex.clear();
+  }
+
+  TEST_F(CurlSession, chunkSegmentedResponse)
+  {
+    // chunked response - simulate the data that the wire will return on every read
+    std::string response0("HTTP/1.1 200 Ok\r");
+    std::string response1("\ntransfer-encoding:");
+    std::string response2(" chunke");
+    std::string response3("d\r\n");
+    std::string response4("\r");
+    std::string response5("\n3\r\n");
+    std::string response6("123");
+    std::string response7("\r\n0\r\n");
+    std::string response8("\r\n");
+
+    std::string connectionKey("connection-key");
+
+    // Can't mock the curMock directly from a unique ptr, heap allocate it first and then make a
+    // unique ptr for it
+    MockCurlNetworkConnection* curlMock = new MockCurlNetworkConnection();
+    EXPECT_CALL(*curlMock, SendBuffer(_, _, _)).WillOnce(Return(CURLE_OK));
+    EXPECT_CALL(*curlMock, ReadFromSocket(_, _, _))
+        .WillOnce(DoAll(
+            SetArrayArgument<1>(response0.data(), response0.data() + response0.size()),
+            Return(response0.size())))
```

Comment:
Help me understand, what is this doing exactly? Why are we  adding the `response0.data()` when passing it to `SetArrayArgument`?

Reply:
This is setting a mocked response for `ReadFromSocket` function. 
The `WillOnce` would define the behavior of the next call to the function. To define it, I use `DoAll`,  this macro takes multiple steps as part of the mock.
First, the `SetArrayArgument<1>` is used to set an output argument from calling `ReadFromSocket`. The second argument (1) will be set, and since it is an array, we need to pass the start and the end address of what we wan to write in the output parameter.  Then the `Return` call will define the return value of `ReadFromSocket`.

This means that, when the test runs, whenever something like `r = ReadFromSocket(arg0, arg1, arg2)` is called, `r` will be assigned to the size of `response0` string and `arg1` will be copied from `response0`.

This gmock is really awesome right :)

I still miss mocka from javascript so much hahha,, but this makes a good job anyway